application-diagram-api/src/main/java/com/xwiki/diagram/internal/PageRenameListener.java

```diff
@@ -0,0 +1,140 @@
+import org.xwiki.job.JobContext;
+import org.xwiki.job.event.JobStartedEvent;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.observation.AbstractEventListener;
+import org.xwiki.observation.ObservationContext;
+import org.xwiki.observation.event.Event;
+
+import com.xpn.xwiki.doc.XWikiDocument;
+
+/**
+ * Listens to rename of pages and starts a thread that will update backlinked diagrams.
+ * 
+ * @version $Id$
+ * @since 1.13
+ */
+@Component
+@Named(PageRenameListener.ROLE_HINT)
+@Singleton
+public class PageRenameListener extends AbstractEventListener implements Disposable
+{
+    /**
+     * The role hint of the document.
+     */
+    protected static final String ROLE_HINT = "PageRenameEventListener";
+
+    /**
+     * Thread that will handle updating diagram's content after a page rename.
+     */
+    public Thread diagramThread;
+
+    @Inject
+    protected ObservationContext observationContext;
+
+    @Inject
+    protected JobContext jobContext;
+
+    @Inject
+    protected ComponentManager componentManager;
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private DiagramRunnable diagramRunnable;
+
+    /**
+     * Constructor.
+     */
+    public PageRenameListener()
+    {
+        super(ROLE_HINT, new DocumentCreatedEvent());
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        if (this.diagramThread == null) {
+            this.diagramThread = new Thread(this.diagramRunnable);
+            this.diagramRunnable.initilizeQueue();
+            this.diagramThread.setName("Update Diagram Links Thread");
+            this.diagramThread.setDaemon(true);
+            this.diagramThread.start();
+        }
+
+        if (observationContext.isIn(new JobStartedEvent("refactoring/rename"))) {
+            Job job = jobContext.getCurrentJob();
+            DocumentReference destinationRef = job.getRequest().getProperty("destination");
+            List<DocumentReference> references = job.getRequest().getProperty("entityReferences");
+
+            if (references != null && references.size() > 0) {
+                XWikiDocument currentDoc = (XWikiDocument) source;
+
+                DocumentReference originalDocRef = references.get(0);
+                DocumentReference currentDocRef = currentDoc.getDocumentReference();
+
+                if (destinationRef.equals(currentDocRef)) {
+                    this.diagramRunnable.addToQueue(new DiagramQueueEntry(originalDocRef, currentDocRef));
+                }
+            }
+        }
+    }
+
+    /**
+     * Actions for closing the thread.
+     * 
+     * @throws InterruptedException if any thread has interrupted the current thread
+     */
+    public void stopUpdateDiagramLinksThread() throws InterruptedException
+    {
+        if (this.diagramThread != null) {
+            this.diagramRunnable.addToQueue(DiagramRunnable.STOP_RUNNABLE_ENTRY);
+            this.diagramThread.interrupt();
```

Comment:
Do you need to interrupt the thread? Isn't it enough to add the stop entry and then wait for it (join)?